### PR TITLE
template: Ensuring template is cross zone if type changed to system

### PR DESCRIPTION
**Pull request by davidjumani**
### Description

This PR enhances https://github.com/apache/cloudstack/pull/3945 by ensuring that all to-be System templates are cross zone

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

### CODE REVIEW

File: server/src/main/java/com/cloud/template/TemplateManagerImpl.java [hunk 1]
```diff
@@ -2198,6 +2202,21 @@ private VMTemplateVO updateTemplateOrIso(BaseUpdateTemplateOrIsoCmd cmd) {
...
+                List<Long> zoneIds = zones.stream().map(zone -> zone.getId()).collect(Collectors.toList());
+                try {
+                    copyTemplate(id, null, zoneIds);
+                } catch (StorageUnavailableException e) {
+                    throw new CloudRuntimeException("Unable to copy template to other zones : ", e);
+                } catch (ResourceAllocationException e) {
+                    throw new CloudRuntimeException("Unable to copy template to other zones : ", e);
+                }
+            }
+        }
```
Comment (DaanHoogland) [inline server/src/main/java/com/cloud/template/TemplateManagerImpl.java:2218]: is a good check but i'd like to see it in a separate method.

**Review by rhtyd — APPROVED** [review]

**Comment by rhtyd** [comment]
cc @ravening @weizhouapache pl review
@blueorangutan package

**Comment by ravening** [comment]
> cc @ravening @weizhouapache pl review
> @blueorangutan package

@rhtyd will review this when its ready

**Comment by davidjumani** [comment]
@blueorangutan package

**Comment by davidjumani** [comment]
@blueorangutan test

File: server/src/main/java/com/cloud/template/TemplateManagerImpl.java [hunk 2]
```diff
@@ -2224,6 +2232,21 @@ else if (details != null && !details.isEmpty()) {
         return _tmpltDao.findById(id);
     }
 
+    private void copyTemplateToAllZones(Long id) {
+        s_logger.info("Copying template to all zones since it will be a system template");
+        List<DataCenterVO> zones = _dcDao.listEnabledZones();
+        if (zones != null) {
+            List<Long> zoneIds = zones.stream().map(zone -> zone.getId()).collect(Collectors.toList());
```
Comment (ravening) [inline server/src/main/java/com/cloud/template/TemplateManagerImpl.java:2239]: Should we skip the zone on which the template is already in?

File: server/src/main/java/com/cloud/template/TemplateManagerImpl.java [hunk 3]
```diff
@@ -2224,6 +2232,21 @@ else if (details != null && !details.isEmpty()) {
         return _tmpltDao.findById(id);
     }
 
+    private void copyTemplateToAllZones(Long id) {
+        s_logger.info("Copying template to all zones since it will be a system template");
```
Comment (ravening) [inline server/src/main/java/com/cloud/template/TemplateManagerImpl.java:2236]: Better to display the template name?

File: server/src/main/java/com/cloud/template/TemplateManagerImpl.java [hunk 4]
```diff
@@ -2224,6 +2232,21 @@ else if (details != null && !details.isEmpty()) {
...
+        List<DataCenterVO> zones = _dcDao.listEnabledZones();
+        if (zones != null) {
+            List<Long> zoneIds = zones.stream().map(zone -> zone.getId()).collect(Collectors.toList());
+            try {
+                copyTemplate(id, null, zoneIds);
+            } catch (StorageUnavailableException e) {
+                throw new CloudRuntimeException("Unable to copy template to other zones : ", e);
```
Comment (ravening) [inline server/src/main/java/com/cloud/template/TemplateManagerImpl.java:2243]: Extract the message to string?

File: server/src/main/java/com/cloud/template/TemplateManagerImpl.java (hunk 2 above)
Comment (weizhouapache) [reply server/src/main/java/com/cloud/template/TemplateManagerImpl.java:2239]: this seems to be handled by https://github.com/apache/cloudstack/blob/c59d0c82137f2294c67d4dd77dd1f4a69b2edb09/server/src/main/java/com/cloud/template/TemplateManagerImpl.java#L931-L936

File: server/src/main/java/com/cloud/template/TemplateManagerImpl.java [hunk 5]
```diff
@@ -2198,6 +2202,10 @@ private VMTemplateVO updateTemplateOrIso(BaseUpdateTemplateOrIsoCmd cmd) {
             template.setDynamicallyScalable(isDynamicallyScalable);
         }
 
+        if (templateType != null && templateType == TemplateType.SYSTEM) {
+            copyTemplateToAllZones(id);
+        }
```
Comment (weizhouapache) [inline server/src/main/java/com/cloud/template/TemplateManagerImpl.java:2207]: @davidjumani should the template be changed to cross-zone when it is done ?

**Comment by davidjumani** [comment]
@weizhouapache @ravening @DaanHoogland Would it be better to make the UpdateTemplate API an async one now that it copies templates between zones ?

File: server/src/main/java/com/cloud/template/TemplateManagerImpl.java (hunk 5 above)
Comment (davidjumani) [reply server/src/main/java/com/cloud/template/TemplateManagerImpl.java:2208]: Makes sense. Added the change

**Comment by davidjumani** [comment]
Updated the PR to just add a cross zone check when trying to change a template to a System / Builtin template

**Comment by davidjumani** [comment]
@blueorangutan package

**Comment by davidjumani** [comment]
@blueorangutan test

**Comment by weizhouapache** [comment]
@davidjumani now it is more clear, good.

**Review by weizhouapache — APPROVED** [review]

**Comment by weizhouapache** [comment]
@davidjumani by the way, if there is only one zone, should it also be ok if it is not cross-zone ?

**Comment by davidjumani** [comment]
@weizhouapache If all or no zones are passed when the template is uploaded, it should work fine even in a single zone scenario

**Comment by weizhouapache** [comment]
> @weizhouapache If 'All zones' or no zoneids are passed when the template is uploaded, it should work fine even in a single zone scenario

@davidjumani yes, the template will be cross-zone. if there is only 1 zone, and upload template for the zone, should it work if change template type to 'SYSTEM' ?

I have another question you may know the answer . Assume there are two zones (zone1/zone2) run with cloudstack 4.14, if register new systemvm template for 4.15 (systemvm-XX-4.15.0) for zone1, will the upgrade to 4.15 work ? will the new systemvm template be copied to zone2 ?

**Comment by davidjumani** [comment]
> > @weizhouapache If 'All zones' or no zoneids are passed when the template is uploaded, it should work fine even in a single zone scenario
> 
> @davidjumani yes, the template will be cross-zone. if there is only 1 zone, and upload template for the zone, should it work if change template type to 'SYSTEM' ?
> 
> I have another question you may know the answer . Assume there are two zones (zone1/zone2) run with cloudstack 4.14, if register new systemvm template for 4.15 (systemvm-XX-4.15.0) for zone1, will the upgrade to 4.15 work ? will the new systemvm template be copied to zone2 ?

@weizhouapache In the first case, I'm not sure it would work, since a zone was selected during the upload. Need to look at the code for a single zone scenario
The second, it will not upgrade since systemvm templates need to be uploaded in all zones that contain the respective hypervisor
@rhtyd Could you confirm ?

**Comment by DaanHoogland** [comment]
I think this addresses the main point that @PaulAngus had with the change template type functionality. He hasn't given a full list indicating there is more, so can we think of more @weizhouapache @davidjumani @rhtyd @ravening ?
This fix looks good to me and should go in before 4.15.

**Comment by PaulAngus** [comment]
> I think this addresses the main point that @PaulAngus had with the change template type functionality. He hasn't given a full list indicating there is more, so can we think of more @weizhouapache @davidjumani @rhtyd @ravening ?
> This fix looks good to me and should go in before 4.15.

**Comment by PaulAngus** [comment]
sorry pressed wrong button. didn't mean to close

**Comment by DaanHoogland** [comment]
:) np, but you did mean to comment I gather

**Comment by PaulAngus** [comment]
yes, the comment was the 'only' intent.  still not blocking, just highlighting/correcting the issues as i see them.

**Comment by DaanHoogland** [comment]
> > I think this addresses the main point that @PaulAngus had with the change template type functionality. He hasn't given a full list indicating there is more, so can we think of more @weizhouapache @davidjumani @rhtyd @ravening ?
> > This fix looks good to me and should go in before 4.15.
> 
> wow, I'm cloudstack QA now?
> I see 'should', 'assume', 'according to code', with no testing.
> There are about 5 different states, which means something like 20 permutations of state change ( as it can't change to itself).

exactly my point, so we need to assess whether for any other of the permutations extra safeguards are needed.

**Comment by rhtyd** [comment]
@davidjumani can you kick pkg/test again?

**Comment by davidjumani** [comment]
@blueorangutan package

**Review by DaanHoogland — APPROVED** [review]

**Comment by davidjumani** [comment]
@blueorangutan test

File: server/src/main/java/com/cloud/template/TemplateManagerImpl.java [hunk 6]
```diff
@@ -2130,7 +2129,7 @@ private VMTemplateVO updateTemplateOrIso(BaseUpdateTemplateOrIsoCmd cmd) {
             return template;
         }
 
-        template = _tmpltDao.createForUpdate(id);
+        template = _tmpltDao.findById(id);
```
Comment (rhtyd) [inline server/src/main/java/com/cloud/template/TemplateManagerImpl.java:2132]: why is this changed?

File: server/src/main/java/com/cloud/template/TemplateManagerImpl.java [hunk 7]
```diff
@@ -2219,7 +2222,7 @@ else if (details != null && !details.isEmpty()) {
             _tmpltDao.saveDetails(template);
         }
 
-        _tmpltDao.update(id, template);
+        _tmpltDao.persist(template);
```
Comment (rhtyd) [inline server/src/main/java/com/cloud/template/TemplateManagerImpl.java:2225]: Here too?

File: server/src/main/java/com/cloud/template/TemplateManagerImpl.java (hunk 6 above)
Comment (davidjumani) [reply server/src/main/java/com/cloud/template/TemplateManagerImpl.java:2132]: createForUpdate creates an empty object with just the id https://github.com/apache/cloudstack/blob/master/framework/db/src/main/java/com/cloud/utils/db/GenericDaoBase.java#L295 , whereas a template object read from the db needed to check whether it's cross zone

File: server/src/main/java/com/cloud/template/TemplateManagerImpl.java (hunk 7 above)
Comment (davidjumani) [reply server/src/main/java/com/cloud/template/TemplateManagerImpl.java:2225]: Required since it's changed from createForUpdate to findById

File: server/src/main/java/com/cloud/template/TemplateManagerImpl.java (hunk 6 above)
Comment (rhtyd) [reply server/src/main/java/com/cloud/template/TemplateManagerImpl.java:2132]: Line 2075 already has that, we need to check why createForUpdate was used in the first place.

File: server/src/main/java/com/cloud/template/TemplateManagerImpl.java (hunk 7 above)
Comment (rhtyd) [reply server/src/main/java/com/cloud/template/TemplateManagerImpl.java:2225]: Persist is usually used for a new record, I'm concerned if this causes any regression or duplicates in DB.

**Comment by rhtyd** [comment]
@davidjumani can you revert the two lines I've comments on, move the conditional up if that makes sense? Let's keep the changes minimal if possible, easier to review and test. Thanks.

File: server/src/main/java/com/cloud/template/TemplateManagerImpl.java (hunk 7 above)
Comment (davidjumani) [reply server/src/main/java/com/cloud/template/TemplateManagerImpl.java:2225]: There are several instances where findById has been used prior to persist. Tested and no duplicate DB entries were created

Comment (rhtyd) [reply server/src/main/java/com/cloud/template/TemplateManagerImpl.java:2225]: I was considering getting this towards 4.15RC2 but the additional changes don't give us confidence - only the conditional check was enough to handle the issue.

Comment (davidjumani) [reply server/src/main/java/com/cloud/template/TemplateManagerImpl.java:2225]: Makes sense, moving the check up

**Comment by davidjumani** [comment]
@blueorangutan package

**Review by rhtyd — APPROVED** [review]

**Comment by rhtyd** [comment]
@weizhouapache @DaanHoogland should we consider this for RC2, with current changes it addresses concerns from #3945
@blueorangutan test

**Comment by weizhouapache** [comment]
> @weizhouapache @DaanHoogland should we consider this for RC2, with current changes it addresses concerns from #3945
> @blueorangutan test

@rhtyd I am ok to merge this into 4.15.0.0
